xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java

```diff
@@ -0,0 +1,169 @@
+
+    RouteMeta(T value, String path, long lastModified, byte[] digest) {
+      this.value = value;
+      this.path = path;
+      this.lastModified = lastModified;
+      this.digest = digest;
+    }
+  }
+
+  private static class RouteConfigFileMetadata {
+    final String path;
+    final long lastModified;
+    final byte[] digest;
+
+    RouteConfigFileMetadata(String path, long lastModified, byte[] digest) {
+      this.path = path;
+      this.lastModified = lastModified;
+      this.digest = digest;
+    }
+  }
+
+  private final ScheduledExecutorService executor;
+  private final Function<String, T> factory;
+  private BiConsumer<T, T> updater;
+  private RouteMeta<T> metadata;
+  private Map<String, RouteConfigFileMetadata> watchFiles =
+      new HashMap<String, RouteConfigFileMetadata>();
+
+  public RouteReloader(ScheduledExecutorService executor, Function<String, T> factory) {
+    this.executor = executor;
+    this.factory = factory;
+    this.updater = null;
+  }
+
+  public T init(String file) {
+    metadata = load(file);
+    addWatchFile(new File(file));
+    return metadata.value;
+  }
+
+  private RouteMeta<T> load(String filePath) {
+    final MessageDigest digest;
+    try {
+      digest = MessageDigest.getInstance("SHA-256");
+    } catch (Exception e) {
+      throw new IllegalStateException("Couldn't load config from file '" + filePath + "'", e);
+    }
+    try {
+      File file = new File(filePath);
+      byte[] encoded = Files.readAllBytes(Paths.get(filePath));
+      String fileContents = new String(encoded, Charset.defaultCharset());
+      T value = factory.apply(fileContents);
+      digest.update(Files.readAllBytes(Paths.get(filePath)));
+      return new RouteMeta<>(value, file.getAbsolutePath(), file.lastModified(), digest.digest());
+    } catch (Exception e) {
+      throw new IllegalStateException("Couldn't load config from file '" + filePath + "'", e);
+    }
+  }
+
+  private boolean shouldPerformUpdate(
+      RouteConfigFileMetadata current, RouteConfigFileMetadata previous) {
+    return (current.lastModified > previous.lastModified
+        && !MessageDigest.isEqual(current.digest, previous.digest));
+  }
+
+  private RouteConfigFileMetadata loadMetaData(File configFile) throws IllegalStateException {
+    final MessageDigest digest;
+    try {
+      digest = MessageDigest.getInstance("SHA-256");
+    } catch (Exception e) {
+      throw new IllegalStateException(
+          "Couldn't load config from file '" + configFile.getAbsolutePath() + "'", e);
+    }
+
+    try {
+      digest.update(Files.readAllBytes(Paths.get(configFile.getAbsolutePath())));
+      return new RouteConfigFileMetadata(
+          configFile.getAbsolutePath(), configFile.lastModified(), digest.digest());
+    } catch (Exception e) {
+      throw new IllegalStateException(
+          "Couldn't load config from file '" + configFile.getAbsolutePath() + "'", e);
+    }
+  }
+
+  private boolean haveWatchFilesChanged() {
```

Comment:
this function is doing more than determining if the watch files have changed - it's also updating the files watched state
maybe a better name would be more appropriate or break apart

Reply:
ill rename it =p